client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala

```diff
@@ -499,62 +563,84 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
             s" shuffleId: $shuffleId $latestLoc")
           return
         }
-        val set = new util.HashSet[RpcCallContext]()
-        set.add(context)
+        val set = new util.HashSet[ChangePartitionRequest]()
+        set.add(changePartition)
         requests.put(partitionId, set)
       }
     }
+  }
 
-    if (cause.isDefined) {
-      blacklistPartition(shuffleId, oldPartition, cause.get)
+  def batchHandleChangePartitions(
+      applicationId: String,
+      shuffleId: Int,
+      changePartitions: Array[ChangePartitionRequest]): Unit = {
+    val requests = changePartitionRequests.get(shuffleId)
+
+    // Blacklist all failed workers
+    if (changePartitions.exists(_.causes.isDefined)) {
+      changePartitions.filter(_.causes.isDefined).foreach { changePartition =>
+        blacklistPartition(shuffleId, changePartition.oldPartition, changePartition.causes.get)
+      }
+    }
+
+    // remove together to reduce lock time
+    def replySuccess(locations: Array[PartitionLocation]): Unit = {
+      requests.synchronized {
+        locations.map { location =>
+          location -> requests.remove(location.getId).asScala.toList
+        }
+      }.foreach { case (newLocation, requests) =>
+        requests.foreach(_.context.reply(ChangeLocationResponse(StatusCode.SUCCESS, newLocation)))
+      }
     }
 
-    def reply(response: ChangeLocationResponse): Unit = {
+    // remove together to reduce lock time
+    def replyFailure(response: ChangeLocationResponse): Unit = {
       requests.synchronized {
-        requests.remove(partitionId)
-      }.asScala.foreach(_.reply(response))
+        changePartitions.flatMap { changePartition =>
+          requests.remove(changePartition.partitionId).asScala.toList
+        }
+      }.foreach(_.context.reply(response))
     }
```

Comment:
Change a lot the the reply method to reduce frequent sync, pls take a look again. @pan3793